### ISSUE TITLE
DAG 10X2 - Pf_Ring support

### ISSUE THREAD

**Issue by dwairlk** [ntop/PF_RING#420]
Hi. 
I have been trying to configure Pf_Ring 7.0 with DAG-5.7.1 version for Ubuntu 16 but pf_ring is not detecting the card.

I  am using pfcount to verify the card.
`pfring_open error [No such device] (pf_ring not loaded or interface dag:0@0 is down ?)`

I have tried these instructions
`sudo pfcount -i dag:0@0`
`sudo pfcount -i dag:dag0:0`
`sudo pfcount -i dag:0:0`
`sudo pfcount -i dag:0`
`sudo pfcount -i dag:dag0`
`sudo pfcount -i dag:0`
`sudo pfcount -i d0`

List the available devices
`sudo pfcount -L`
```Available devices (-i):

 0. lo
 1. enp10s0
 2. dag:0@0
 3. dag:0@2
 4. dag:0@4
 5. dag:0@6
 6. dag:0@8
 7. dag:0@10
```

For the normal network card, testing pf_ring
`sudo pfcount -i enp10s0`
```\Using PF_RING v.7.3.0
Capturing from enp10s0 [mac: XX:XX:XX][if_index: 2][speed: 100Mb/s]
# Device RX channels: 1
# Polling threads:    1
Dumping statistics on /proc/net/pf_ring/stats/27040-enp10s0.22
=========================

Absolute Stats: [5 pkts total][0 pkts dropped][0.0% dropped]
[5 pkts rcvd][696 bytes rcvd]
```

Testing DAG card
`sudo dagconfig -d0 -u`

```**** Number of blocks: 1
Nb of counter(s) in Block ID "Framer_Tx_Rx": 22
Port:  A
              Rx_Bytes: 83254254
             Rx_Frames: 98935```


Some configurations:
`sudo daginf

card          dag0
id            0
model         DAG 10X2-P
device        0xa120
board rev     0
iom size      65536 (64kB)
bus id        0000:06:00.0
dev NUMA node 0
rxstreams     32
txstreams     1`


`modinfo pf_ring && cat /proc/net/pf_ring/info
filename:       /lib/modules/4.4.0-131-generic/kernel/net/pf_ring/pf_ring.ko
alias:          net-pf-27
version:        7.3.0
description:    Packet capture acceleration and analysis
author:         ntop.org
license:        GPL
srcversion:     88FE8EFC5100E3EB91C3EF0
depends:
retpoline:      Y
vermagic:       4.4.0-131-generic SMP mod_unload modversions retpoline
parm:           min_num_slots:Min number of ring slots (uint)
parm:           perfect_rules_hash_size:Perfect rules hash size (uint)
parm:           enable_tx_capture:Set to 1 to capture outgoing packets (uint)
parm:           enable_frag_coherence:Set to 1 to handle fragments (flow coherence) in clusters (uint)
parm:           enable_ip_defrag:Set to 1 to enable IP defragmentation(only rx traffic is defragmentead) (uint)
parm:           quick_mode:Set to 1 to run at full speed but with upto one socket per interface (uint)
parm:           force_ring_lock:Set to 1 to force ring locking (automatically enable with rss) (uint)
parm:           enable_debug:Set to 1 to enable PF_RING debug tracing into the syslog, 2 for more verbosity (uint)
parm:           transparent_mode:(deprecated) (uint)
PF_RING Version          : 7.3.0 (dev:c85efbc90d5abb7ef471be17cf9192b88a842ac4)
Total rings              : 0

Standard (non ZC) Options
Ring slots               : 4096
Slot version             : 17
Capture TX               : Yes [RX+TX]
IP Defragment            : No
Socket Mode              : Standard
Cluster Fragment Queue   : 0
Cluster Fragment Discard : 0`


I installed all the .deb files for Ubuntu 16.04 provided by Endace
`sudo apt-get install ./*.deb`
then the module was loaded
`sudo dagload`

Pf_ring:
`cd PF_RING/kernel; make; sudo make install; sudo insmod pf_ring.ko`

Endace:

`cd ../userland/modules/Endace`

`./configure
checking for gcc... gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables...
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether gcc accepts -g... yes
checking for gcc option to accept ISO C89... none needed
checking for dag_get_stream_erf_types in -ldag... yes
checking if dag is present... yes
configure: creating ./config.status
config.status: creating Makefile`

`make
gcc  -O2  -Wall -fPIC -I../../../kernel -I../../nbpf -I../../lib  -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=0 -c pfring_mod_dag.c -o pfring_mod_dag.o
=*= making library ../../lib/libs/libpfring_dag_x86_64.a =*=
ar rs ../../lib/libs/libpfring_dag_x86_64.a pfring_mod_dag.o pfring_mod_dag.h
ranlib ../../lib/libs/libpfring_dag_x86_64.a
=*= making library ../../lib/libs/libpfring_dag_x86_64.so =*=
gcc  -O2  -shared pfring_mod_dag.o -lpthread -o ../../lib/libs/libpfring_dag_x86_64.so`

`sudo make install
make clean; make GCCARCH="corei7"
make[1]: Entering directory '/PF_RING/userland/modules/Endace'
make[1]: Leaving directory '/PF_RING/userland/modules/Endace'
make[1]: Entering directory '/PF_RING/userland/modules/Endace'
gcc -march=corei7 -mtune=corei7 -O2  -Wall -fPIC -I../../../kernel -I../../nbpf -I../../lib  -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=0 -c pfring_mod_dag.c -o pfring_mod_dag.o
=*= making library ../../lib/libs/libpfring_dag_x86_64_corei7.a =*=
ar rs ../../lib/libs/libpfring_dag_x86_64_corei7.a pfring_mod_dag.o pfring_mod_dag.h
ranlib ../../lib/libs/libpfring_dag_x86_64_corei7.a
=*= making library ../../lib/libs/libpfring_dag_x86_64.so =*=
gcc -march=corei7 -mtune=corei7 -O2  -shared pfring_mod_dag.o -lpthread -o ../../lib/libs/libpfring_dag_x86_64.so
make[1]: Leaving directory '/PF_RING/userland/modules/Endace'
make clean; make GCCARCH="corei7-avx"
make[1]: Entering directory '/PF_RING/userland/modules/Endace'
make[1]: Leaving directory '/PF_RING/userland/modules/Endace'
make[1]: Entering directory '/PF_RING/userland/modules/Endace'
gcc -march=corei7-avx -mtune=corei7-avx -O2  -Wall -fPIC -I../../../kernel -I../../nbpf -I../../lib  -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=0 -c pfring_mod_dag.c -o pfring_mod_dag.o
=*= making library ../../lib/libs/libpfring_dag_x86_64_corei7-avx.a =*=
ar rs ../../lib/libs/libpfring_dag_x86_64_corei7-avx.a pfring_mod_dag.o pfring_mod_dag.h
ranlib ../../lib/libs/libpfring_dag_x86_64_corei7-avx.a
=*= making library ../../lib/libs/libpfring_dag_x86_64.so =*=
gcc -march=corei7-avx -mtune=corei7-avx -O2  -shared pfring_mod_dag.o -lpthread -o ../../lib/libs/libpfring_dag_x86_64.so
make[1]: Leaving directory '/PF_RING/userland/modules/Endace'
make clean; make GCCARCH="core-avx2"
make[1]: Entering directory '/PF_RING/userland/modules/Endace'
make[1]: Leaving directory '/PF_RING/userland/modules/Endace'
make[1]: Entering directory '/PF_RING/userland/modules/Endace'
gcc -march=core-avx2 -mtune=core-avx2 -O2  -Wall -fPIC -I../../../kernel -I../../nbpf -I../../lib  -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=0 -c pfring_mod_dag.c -o pfring_mod_dag.o
=*= making library ../../lib/libs/libpfring_dag_x86_64_core-avx2.a =*=
ar rs ../../lib/libs/libpfring_dag_x86_64_core-avx2.a pfring_mod_dag.o pfring_mod_dag.h
ranlib ../../lib/libs/libpfring_dag_x86_64_core-avx2.a
=*= making library ../../lib/libs/libpfring_dag_x86_64.so =*=
gcc -march=core-avx2 -mtune=core-avx2 -O2  -shared pfring_mod_dag.o -lpthread -o ../../lib/libs/libpfring_dag_x86_64.so
make[1]: Leaving directory '/PF_RING/userland/modules/Endace'
make clean; make GCCARCH=""
make[1]: Entering directory '/PF_RING/userland/modules/Endace'
make[1]: Leaving directory '/PF_RING/userland/modules/Endace'
make[1]: Entering directory '/PF_RING/userland/modules/Endace'
gcc  -O2  -Wall -fPIC -I../../../kernel -I../../nbpf -I../../lib  -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=0 -c pfring_mod_dag.c -o pfring_mod_dag.o
=*= making library ../../lib/libs/libpfring_dag_x86_64.a =*=
ar rs ../../lib/libs/libpfring_dag_x86_64.a pfring_mod_dag.o pfring_mod_dag.h
ranlib ../../lib/libs/libpfring_dag_x86_64.a
=*= making library ../../lib/libs/libpfring_dag_x86_64.so =*=
gcc  -O2  -shared pfring_mod_dag.o -lpthread -o ../../lib/libs/libpfring_dag_x86_64.so
make[1]: Leaving directory '/PF_RING/userland/modules/Endace'`

then,
`cd ../../lib; ./configure && make && sudo make install`

`cd ../libpcap; ./configure && make && sudo make install`

finally,
`cd ../examples; make && sudo make install`


Is this version of DAG supported? Is there any additional compilation or configuration?

**Comment by cardigliano** [comment]
Hi @dwairlk, please try following the instructions at http://www.ntop.org/guides/pf_ring/modules/dag.html including recompiling the 'Endace' module. Those instructions are working in our lab using dag-5.5, they are supposed to work with 5.7.

**Comment by dwairlk** [comment]
Hi @cardigliano,
Endace has not released the source for the version 5.7. So I compiled version 5.6.0, as follows:
1. Install the required packages from the Ubuntu repository.
`sudo apt install flex bison byacc make 
  gcc libhwloc-dev libxml2  libxml2-dev \
linux-headers-generic  pkg-config 
python-dev dkms `

2. Install libntar from the source
`tar -vxzf libntar-1.1.0.tar.gz; cd libntar-1.1.0/;  ./configure && make && sudo make install`

3. Install libpcap from the source
`tar -vxzf libpcap-1.7.4.tgz; cd libpcap-1.7.4.tgz/;  ./configure && make && sudo make install`

4. DAG SDK
`tar -xvzf dag-5.6.0.tar.gz; cd dag-5.6.0/; ./configure && make && make install`

5. Follow the steps at http://www.ntop.org/guides/pf_ring/modules/dag.html

It is working.
`sudo pfcount -i dag:0@0
Using PF_RING v.7.3.0
Capturing from dag:0@0 [mac: unknown][if_index: -1][speed: 0Mb/s]`
`# Device RX channels: 1`
`# Polling threads:    1`
`=========================`
`Absolute Stats: [0 pkts total][0 pkts dropped][0.0% dropped]
[0 pkts rcvd][0 bytes rcvd]`
`=========================`

Thank you @cardigliano

**Comment by sfd** [comment]
This issue has now been fixed in https://github.com/ntop/PF_RING/pull/464.
You should be able to install DAG 5.7.1 (or other releases) from the binary Deb packages in Ubuntu, and the 'dev' branch of PF_RING on Github will now find the DAG libraries correctly. There is no need to recompile DAG from sources.
